### DISCUSSION TITLE
S3 Access from Sagemaker: DuckDB slower than expected... threading?

### DISCUSSION THREAD

**Discussion by CerebralMastication** [duckdb/duckdb#4780] [Q&A]
Hey all, I've been playing around with DuckDB on Sagemaker reading parquet files on S3. I expected it to be considerably slower than a local file system, of course. However the performance has been really surprisingly slow. Slow enough that I'm wondering if I'm doing something wrong or if possibly DuckDB isn't reading from S3 in parallel threads the way it does on a local file system. 

I have a bucket in S3 with a ~7000 parquet files. I'm using only the first 101 of these for my testing. The script that generates the test data can be found [here](https://github.com/CerebralMastication/DataToolTesting/blob/main/simulate_imbalanced_data.ipynb). 

my test of reading from S3 with duckdb is in [this notebook](https://github.com/CerebralMastication/DataToolTesting/blob/main/query_s3.ipynb). 

The crux of the test is this simple summation:

```
df = con.execute(f"select sum(value) from parquet_scan({my_parquet_list[0:100]}) ;").df()
```

I'm passing `parquet_scan` a list of 101 files which represent approx 165MB and 11.5m records. What's unexpected is I ran this test using 4 different Sagemaker instance sizes with 2, 4, 8, and 16 VCPU. They all ran the query in ~12 seconds. (I used `%%timeit` cell magic so ran 7 loops and I reran each test 3 times. I found some variation around run time, but it was not related to number of CPUs which I found unexpected as I would have expected read speed to decrease with increasing number of vCPUs. 

I'm also a bit surprised that this read took 12 seconds even on a single thread. With my data and compute both located in `eu-west-1 ` I would have expected to get 100 Gbps throughput. 

It feels like maybe I'm doing this wrong. Any suggestions you all have would be greatly appreciated. 

Update:
I wanted to compare DuckDB timing above with simply slurping all the parquet files into Pandas and filtering them. I realize that this is RAM constrained but I know that the `awswrangler` is single threaded so that will give me some notion of how fast simply reading the data from S3 on one thread is. So I ran the following:

```
df = wr.s3.read_parquet(path=my_parquet_list[0:100], boto3_session=my_session)
df.value.sum()
```
which should move all those same parquet files out of S3 and read them into ram then filter. 

it runs in 8.5 seconds on a worker with 4 vCPU and 16GiB RAM (ml.g4dn.xlarge)

**Comment by hannes** [comment]
Maybe @samansmink can comment on this?

**Comment by samansmink** [comment]
Hey @CerebralMastication thanks for the detailed description. Increasing the number of CPUs available should definitely increase the speed of your query. 

I did a quick check with some data I had on S3 but there is a clear speedup there when using a different numbers of threads/CPUs. I will use your scripts later today to see if I can reproduce the behaviour and figure out whats happening.

What most definitely helps is to increase the number of threads in DuckDB to something like ~2x the amount of cpus available for S3 workloads, e.g.: 
```sql
PRAGMA threads=16;
```
Since DuckDB by default only launches as many threads as there are CPUs, giving a few more will reduce time spent waiting for I/O.

**Reply by CerebralMastication** [reply]
I’ll test forcing the threads up and see what happens. 

Is there a diagnostic I can print that will tell me how many threads DuckDB is using?

**Reply by CerebralMastication** [reply]
here's two runs. 

first on a machine with 2 vCPU (4GiB RAM):

![image](https://user-images.githubusercontent.com/126879/192105172-c74a141f-9e18-420e-be78-3c95120a6260.png)

that looks like it shows speed up from 1 to 2 threads. But then it's no faster with more threads, which makes sense with 2 vCPU. 

The second run is on a machine with 8 vCPU (32 GiB RAM):
(the top run, not shown in the first image is the default behavior with no `PRAGMA threads` statement

![image](https://user-images.githubusercontent.com/126879/192105838-4ab6a0fb-ff94-4cec-a5f3-c64c0292334e.png)

While the single thread run is slower than the 2 thread run, on average, the fastest single thread run is faster than the slowest 2 thread run (hand waving away the methodological approach of `%%timeit` )

but more importantly, the 4 thread and 8 thread runs are not distinguishable from the 2 thread run. I find that surprising.

**Reply by CerebralMastication** [reply]
refactored slightly to show the actual run times of all 7 loops for each `threads` setting. 

![image](https://user-images.githubusercontent.com/126879/192107126-43ac1594-8500-4986-a07d-d479b23501ba.png)

this is on the 8 vCPU and 32GiB instance

**Comment by samansmink** [comment]
@CerebralMastication There's a PR up (https://github.com/duckdb/duckdb/pull/5405) that should bring significant improvements to scanning many parquet files over HTTP. There were indeed some issues that caused parts of the scanning process to be basically single threaded! Thanks for your input here 👍

**Comment by djouallah** [comment]
@samansmink are those improvement will help with arrow dataset when Querying Object store like GCP ?

**Comment by samansmink** [comment]
@djouallah I don't think so, the parquet reader changes are specific to the parquet reader. The http metadata cache will only reduce the number of head requests made by duckdb. This is only for http requests that duckdb itself makes through the httpfs, which is not the case when reading an arrow dataset currently.
